File: uploadservice/src/main/java/net/gotev/uploadservice/http/impl/HurlStackConnection.java
package net.gotev.uploadservice.http.impl;

import net.gotev.uploadservice.Logger;
import net.gotev.uploadservice.NameValue;
import net.gotev.uploadservice.ServerResponse;
import net.gotev.uploadservice.UploadService;
import net.gotev.uploadservice.http.HttpConnection;

import java.io.ByteArrayOutputStream;
import java.io.IOException;
import java.io.InputStream;
import java.net.HttpURLConnection;
import java.net.URL;
import java.util.LinkedHashMap;
import java.util.List;
import java.util.Map;

import javax.net.ssl.HttpsURLConnection;

/**
 * {@link HttpConnection} implementation using {@link HttpURLConnection}.
 * @author gotev (Aleksandar Gotev)
 */
public class HurlStackConnection implements HttpConnection {

    private static final String LOG_TAG = HurlStackConnection.class.getSimpleName();

    private HttpURLConnection mConnection;

    public HurlStackConnection(String method, String url, boolean followRedirects,
                               boolean useCaches, int connectTimeout, int readTimeout)
            throws IOException {
        Logger.debug(getClass().getSimpleName(), "creating new connection");

        URL urlObj = new URL(url);

        if (urlObj.getProtocol().equals("https")) {
            mConnection = (HttpsURLConnection) urlObj.openConnection();
        } else {
            mConnection = (HttpURLConnection) urlObj.openConnection();
        }

        mConnection.setDoInput(true);
        mConnection.setDoOutput(true);
        mConnection.setConnectTimeout(connectTimeout);
        mConnection.setReadTimeout(readTimeout);
        mConnection.setUseCaches(useCaches);
        mConnection.setInstanceFollowRedirects(followRedirects);
        mConnection.setRequestMethod(method);
    }

    @Override
    public HttpConnection setHeaders(List<NameValue> requestHeaders) throws IOException {
        for (final NameValue param : requestHeaders) {
            mConnection.setRequestProperty(param.getName(), param.getValue());
        }

        return this;
    }

    @Override
    public HttpConnection setTotalBodyBytes(long totalBodyBytes, boolean isFixedLengthStreamingMode) {
        if (isFixedLengthStreamingMode) {
            if (android.os.Build.VERSION.SDK_INT >= 19) {
                mConnection.setFixedLengthStreamingMode(totalBodyBytes);

            } else {
                if (totalBodyBytes > Integer.MAX_VALUE)
                    throw new RuntimeException("You need Android API version 19 or newer to "
                            + "upload more than 2GB in a single request using "
                            + "fixed size content length. Try switching to "
                            + "chunked mode instead, but make sure your server side supports it!");

                mConnection.setFixedLengthStreamingMode((int) totalBodyBytes);
            }
        } else {
            mConnection.setChunkedStreamingMode(0);
        }

        return this;
    }

    private byte[] getServerResponseBody() throws IOException {
        InputStream stream = null;

        try {
            if (mConnection.getResponseCode() / 100 == 2) {
                stream = mConnection.getInputStream();
            } else {
                stream = mConnection.getErrorStream();
            }

            return getResponseBodyAsByteArray(stream);

        } finally {
            if (stream != null) {
                try {
                    stream.close();
                } catch (Exception exc) {
                    Logger.error(LOG_TAG, "Error while closing server response stream", exc);
                }
            }
        }
    }

    private byte[] getResponseBodyAsByteArray(final InputStream inputStream) {
        ByteArrayOutputStream byteStream = new ByteArrayOutputStream();

        byte[] buffer = new byte[UploadService.BUFFER_SIZE];
        int bytesRead;

        try {
            while ((bytesRead = inputStream.read(buffer, 0, buffer.length)) > 0) {
                byteStream.write(buffer, 0, bytesRead);
            }
        } catch (Exception ignored) {}

        return byteStream.toByteArray();
    }

    private LinkedHashMap<String, String> getServerResponseHeaders() throws IOException {
        Map<String, List<String>> headers = mConnection.getHeaderFields();
        if (headers == null)
            return null;

        LinkedHashMap<String, String> out = new LinkedHashMap<>(headers.size());

        for (Map.Entry<String, List<String>> entry : headers.entrySet()) {
            if (entry.getKey() != null) {
                StringBuilder headerValue = new StringBuilder();
                for (String value : entry.getValue()) {
                    headerValue.append(value);
                }
                out.put(entry.getKey(), headerValue.toString());
            }
        }

        return out;
    }

    @Override
    public ServerResponse getResponse(final RequestBodyDelegate delegate) throws IOException {

        final HurlBodyWriter bodyWriter = new HurlBodyWriter(mConnection.getOutputStream());
        delegate.onBodyReady(bodyWriter);
        bodyWriter.flush();

        return new ServerResponse(mConnection.getResponseCode(),
                getServerResponseBody(), getServerResponseHeaders());
    }

    @Override
    public void close() {
        Logger.debug(getClass().getSimpleName(), "closing connection");

        if (mConnection != null) {
            try {
                mConnection.getInputStream().close();
            } catch (Exception ignored) { }

            try {
                mConnection.getOutputStream().flush();
                mConnection.getOutputStream().close();
            } catch (Exception ignored) { }

            try {
                mConnection.disconnect();
            } catch (Exception exc) {
                Logger.error(LOG_TAG, "Error while closing connection", exc);
            }
        }
    }
}

File: uploadservice/src/main/java/net/gotev/uploadservice/network/hurl/HurlStackConnection.kt
package net.gotev.uploadservice.network.hurl

import net.gotev.uploadservice.logger.UploadServiceLogger
import net.gotev.uploadservice.NameValue
import net.gotev.uploadservice.network.ServerResponse
import net.gotev.uploadservice.network.HttpConnection
import java.io.IOException
import java.net.HttpURLConnection
import java.net.URL
import java.util.*
import javax.net.ssl.HttpsURLConnection
import kotlin.collections.LinkedHashMap

/**
 * [HttpConnection] implementation using [HttpURLConnection].
 * @author gotev (Aleksandar Gotev)
 */
class HurlStackConnection(
        method: String,
        url: String,
        followRedirects: Boolean,
        useCaches: Boolean,
        connectTimeout: Int,
        readTimeout: Int) : HttpConnection {

    private val connection: HttpURLConnection
    private val uuid = UUID.randomUUID().toString()

    private fun String.createConnection(): HttpURLConnection {
        val url = URL(trim())

        return if (url.protocol.toLowerCase() == "https") {
            url.openConnection() as HttpsURLConnection
        } else {
            url.openConnection() as HttpURLConnection
        }
    }

    init {
        UploadServiceLogger.debug(javaClass.simpleName, "creating new HttpURLConnection (uuid: $uuid)")

        connection = url.createConnection().apply {
            doInput = true
            doOutput = true
            this.connectTimeout = connectTimeout
            this.readTimeout = readTimeout
            this.useCaches = useCaches
            instanceFollowRedirects = followRedirects
            requestMethod = method
        }
    }

    private val responseBody: ByteArray
        @Throws(IOException::class)
        get() {
            return if (connection.responseCode / 100 == 2) {
                connection.inputStream
            } else {
                connection.errorStream
            }.use {
                it.readBytes()
            }
        }

    private val responseHeaders: LinkedHashMap<String, String>
        @Throws(IOException::class)
        get() {
            val headers = connection.headerFields ?: return LinkedHashMap(0)

            return LinkedHashMap<String, String>(headers.size).apply {
                headers.entries
                        .filter { it.key != null && it.value != null && it.value.isNotEmpty() }
                        .forEach { (key, values) ->
                            this[key] = values.first()
                        }
            }
        }

    @Throws(IOException::class)
    override fun setHeaders(requestHeaders: List<NameValue>): HttpConnection {
        for (param in requestHeaders) {
            connection.setRequestProperty(param.name.trim(), param.value.trim())
        }

        return this
    }

    override fun setTotalBodyBytes(totalBodyBytes: Long, isFixedLengthStreamingMode: Boolean): HttpConnection {
        connection.apply {
            if (isFixedLengthStreamingMode) {
                setFixedLengthStreamingMode(totalBodyBytes)
            } else {
                setChunkedStreamingMode(0)
            }
        }

        return this
    }

    @Throws(IOException::class)
    override fun getResponse(delegate: HttpConnection.RequestBodyDelegate): ServerResponse {
        HurlBodyWriter(connection.outputStream).use {
            delegate.onWriteRequestBody(it)
        }

        return ServerResponse(connection.responseCode, responseBody, responseHeaders)
    }

    override fun close() {
        UploadServiceLogger.debug(javaClass.simpleName, "closing HttpURLConnection (uuid: $uuid)")

        try {
            connection.inputStream.close()
        } catch (ignored: Throwable) {
        }

        try {
            connection.outputStream.flush()
        } catch (ignored: Throwable) {
        }

        try {
            connection.outputStream.close()
        } catch (ignored: Throwable) {
        }

        try {
            connection.disconnect()
        } catch (ignored: Throwable) {
        }
    }
}
